packages/flutter/lib/src/material/page.dart

```diff
@@ -172,7 +200,12 @@ class MaterialPage<T> extends Page<T> {
 
   @override
   Route<T> createRoute(BuildContext context) {
-    return _PageBasedMaterialPageRoute<T>(page: this, allowSnapshotting: allowSnapshotting);
+    final _PageBasedMaterialPageRoute<T> route = _PageBasedMaterialPageRoute<T>(page: this, allowSnapshotting: allowSnapshotting);
+    // It is necessary to do this now as this is the earliest the MaterialPage can
+    // know what its context is for deciding what delegated transition to send.
+    // buildTransitions happens after canTransitionTo in a lot of cases.
```

Comment:
Does that mean`MaterialRouteTransitionMixin` can't catch some of the case?

Reply:
It could be aware of these cases, but not do anything about it in time.

Reply:
That seems to be an issue. This method is only used in MaterialPage but not MaterialPageRoute. That means if developer does not use page API this could break.

Reply:
The MaterialPageRoute case is covered by didChangeNext in the base ModalRoute class. Setting it here for MaterialPage is not strictly necessary, I believe, but without it there's a chance of unnecessary rebuilds when using the page API. I could update this comment to explain this better.